cronet/src/main/java/io/grpc/cronet/CronetCallOptions.java

```diff
@@ -27,8 +29,34 @@ private CronetCallOptions() {}
    * get Cronet metrics from {@link org.chromium.net.RequestFinishedInfo.Listener} with the same
    * annotation object.
    *
-   * The Object must not be null.
+   * <p>The Object must not be null.
+   *
+   * @deprecated Use {@link CronetCallOptions#withAnnotation} instead.
    */
-    public static final CallOptions.Key<Object> CRONET_ANNOTATION_KEY =
+  @Deprecated
+  public static final CallOptions.Key<Object> CRONET_ANNOTATION_KEY =
       CallOptions.Key.of("cronet-annotation", null);
+
+  /**
+   * Returns a copy of {@code callOptions} with {@code annotation} included in the set of Cronet
+   * annotation objects. When an RPC is made using a {@link CallOptions} instance returned by this
+   * method, the annotation objects will be attached to the underlying Cronet bidirectional stream.
+   * When the stream finishes, the user can retrieve the annotation object(s) via {@link
+   * org.chromium.net.RequestFinishedInfo.Listener}.
+   *
+   * @param annotation the object to attach to the Cronet stream
+   */
+  public static CallOptions withAnnotation(CallOptions callOptions, Object annotation) {
+    Set<Object> annotations = callOptions.getOption(CRONET_ANNOTATIONS_KEY);
+    if (annotations != null) {
```

Comment:
```java
if (annotations == null) {
  annotations = new HashSet<Object>();
}
annotations.add(annotation);
```

However, since options are mostly immutable, it may be strange this modifies the set. Notice that the `withOption` at the end isn't necessary if the key was already present.

Reply:
Switched to Collection backed by ArrayList (also matches how [Cronet handles the annotations](https://cs.chromium.org/chromium/src/components/cronet/android/java/src/org/chromium/net/impl/BidirectionalStreamBuilderImpl.java?type=cs&q=addRequestAnnotation&l=123)) and recreate a new ArrayList on each call. I think this better respects the immutability of call options; is this type of change what you had in mind?

Reply:
Yeah, that's what I was thinking.